rmw/include/rmw/rmw.h

```diff
@@ -2787,6 +2788,127 @@ RMW_WARN_UNUSED
 rmw_ret_t
 rmw_set_log_severity(rmw_log_severity_t severity);
 
+/// Set callback function of the rmw subscription listener.
+/**
+ * This API sets the callback function which will be called whenever the
+ * subscription listener is notified about a new message for the subscription.
+ * The callback will be called from background thread(s) that the rmw
+ * implementation uses to wait (listen) for events to happen.
```

Comment:
Technically we cannot enforce that, nor does it make sense to require it. I'd say instead:

```suggestion
 * The callback may be called from a thread that the rmw implementation
 * created, rather than a thread owned by the user, i.e. some thread other
 * than user owned threads calling rmw functions such as rmw_wait() or
 * rmw_publish().
```

Basically, we should let the user know that what thread this comes from is not known, and in this way we do allow implementations to do what ever they need in order to implement this. For example, they may choose to call directly from within the `rmw_wait()` function in some cases or perhaps call subscription callbacks directly from `rmw_publish()`. I don't know if this advisable, but rather I would not want to box in what the rmw can do in the documentation of this function.

This is not mutually exclusive (in my opinion) to documenting that these callbacks must be serviced even if `rmw_wait()` (or any other rmw function) is never called.

What do you guys think? @ivanpauno

Reply:
That sounds good to me